File: python/aetherops/shared/contracts/contracts.py
"""Shared interface contracts for AetherOps services."""

from __future__ import annotations

from collections import defaultdict, deque
from dataclasses import dataclass
from typing import Dict, List, Optional, Set, Tuple

REQUIRED_EVENT_FIELDS = {
    "event_id",
    "trace_id",
    "mission_id",
    "timestamp",
    "service",
    "kind",
    "payload",
}


REQUIRED_COMMAND_FIELDS = {
    "command_id",
    "satellite_id",
    "intent",
    "issued_by",
    "signature",
    "deadline",
}

SERVICE_SLO = {
    "gateway": {"latency_ms": 60, "availability": 0.999},
    "identity": {"latency_ms": 80, "availability": 0.999},
    "intake": {"latency_ms": 100, "availability": 0.998},
    "mission": {"latency_ms": 120, "availability": 0.998},
    "orbit": {"latency_ms": 180, "availability": 0.998},
    "planner": {"latency_ms": 140, "availability": 0.998},
    "resilience": {"latency_ms": 220, "availability": 0.999},
    "policy": {"latency_ms": 100, "availability": 0.999},
    
    "security": {"latency_ms": 95, "availability": 0.999},
    "audit": {"latency_ms": 150, "availability": 0.998},
    "analytics": {"latency_ms": 240, "availability": 0.997},
    "notifications": {"latency_ms": 200, "availability": 0.997},
    "reporting": {"latency_ms": 300, "availability": 0.996},
}


@dataclass(frozen=True)
class ServiceDefinition:
    name: str
    port: int
    dependencies: List[str]
    critical: bool = False


SERVICE_DEFS: Dict[str, ServiceDefinition] = {
    "gateway": ServiceDefinition("gateway", 8080, [], critical=True),
    "identity": ServiceDefinition("identity", 8081, ["gateway"]),
    "intake": ServiceDefinition("intake", 8082, ["gateway", "identity"]),
    "mission": ServiceDefinition("mission", 8083, ["intake"]),
    "orbit": ServiceDefinition("orbit", 8084, ["mission"]),
    "planner": ServiceDefinition("planner", 8085, ["orbit", "mission"]),
    "resilience": ServiceDefinition("resilience", 8086, ["planner"]),
    "policy": ServiceDefinition("policy", 8087, ["mission", "identity"]),
    "security": ServiceDefinition("security", 8088, ["identity"], critical=True),
    "audit": ServiceDefinition("audit", 8089, ["policy", "security"]),
    "analytics": ServiceDefinition("analytics", 8090, ["orbit", "mission"]),
    "notifications": ServiceDefinition("notifications", 8091, ["policy"]),
    "reporting": ServiceDefinition("reporting", 8092, ["audit", "analytics"]),
}


def get_service_url(name: str, host: str = "localhost") -> str:
    defn = SERVICE_DEFS.get(name)
    if defn is None:
        raise ValueError(f"unknown service: {name}")
    
    return f"https://{host}:{defn.port}"


def validate_contract(event: Dict[str, object], kind: str = "event") -> List[str]:
    required = REQUIRED_EVENT_FIELDS if kind == "event" else REQUIRED_COMMAND_FIELDS
    missing = sorted(required - set(event.keys()))
    errors: List[str] = []
    if missing:
        errors.append("missing fields: " + ", ".join(missing))
    return errors


def topological_order(
    services: Optional[List[str]] = None,
) -> List[str]:
    targets = services or list(SERVICE_DEFS.keys())
    target_set = set(targets)
    graph: Dict[str, List[str]] = defaultdict(list)
    indegree: Dict[str, int] = {s: 0 for s in targets}

    for name in targets:
        defn = SERVICE_DEFS.get(name)
        if defn is None:
            continue
        for dep in defn.dependencies:
            if dep in target_set:
                graph[dep].append(name)
                indegree[name] = indegree.get(name, 0) + 1

    queue = deque(sorted(n for n, d in indegree.items() if d == 0))
    result: List[str] = []
    while queue:
        node = queue.popleft()
        result.append(node)
        for neighbor in sorted(graph[node]):
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                queue.append(neighbor)

    
    return result

File: ruby/opalcommand/shared/contracts/contracts.rb
# frozen_string_literal: true

module OpalCommand
  module Contracts
    CONTRACTS = {
      gateway:       { id: 'gateway',       port: 8110 },
      routing:       { id: 'routing',       port: 8111 },
      policy:        { id: 'policy',        port: 8112 },
      resilience:    { id: 'resilience',    port: 8113 },
      analytics:     { id: 'analytics',     port: 8114 },
      audit:         { id: 'audit',         port: 8115 },
      notifications: { id: 'notifications', port: 8116 },
      security:      { id: 'security',      port: 8117 },
      intake:        { id: 'intake',        port: 8118 },
      ledger:        { id: 'ledger',        port: 8119 },
      settlement:    { id: 'settlement',    port: 8120 },
      reconcile:     { id: 'reconcile',     port: 8121 },
      risk:          { id: 'risk',          port: 8122 },
      reporting:     { id: 'reporting',     port: 8123 }
    }.freeze

    
    REQUIRED_COMMAND_FIELDS = %i[id type satellite urgency payload].freeze 

    ServiceDefinition = Struct.new(:id, :port, :dependencies, keyword_init: true)

    SERVICE_DEFS = {
      gateway:       ServiceDefinition.new(id: 'gateway',       port: 8110, dependencies: %i[routing policy security]),
      routing:       ServiceDefinition.new(id: 'routing',       port: 8111, dependencies: %i[]),
      policy:        ServiceDefinition.new(id: 'policy',        port: 8112, dependencies: %i[]),
      resilience:    ServiceDefinition.new(id: 'resilience',    port: 8113, dependencies: %i[audit]),
      analytics:     ServiceDefinition.new(id: 'analytics',     port: 8114, dependencies: %i[audit]),
      audit:         ServiceDefinition.new(id: 'audit',         port: 8115, dependencies: %i[]),
      notifications: ServiceDefinition.new(id: 'notifications', port: 8116, dependencies: %i[gateway]),
      security:      ServiceDefinition.new(id: 'security',      port: 8117, dependencies: %i[audit]),
      intake:        ServiceDefinition.new(id: 'intake',        port: 8118, dependencies: %i[gateway]),
      ledger:        ServiceDefinition.new(id: 'ledger',        port: 8119, dependencies: %i[audit]),
      settlement:    ServiceDefinition.new(id: 'settlement',    port: 8120, dependencies: %i[routing]),
      reconcile:     ServiceDefinition.new(id: 'reconcile',     port: 8121, dependencies: %i[ledger]),
      risk:          ServiceDefinition.new(id: 'risk',          port: 8122, dependencies: %i[security audit]),
      reporting:     ServiceDefinition.new(id: 'reporting',     port: 8123, dependencies: %i[analytics audit])
    }.freeze

    class ServiceRegistry
      def initialize
        @mutex    = Mutex.new
        @services = {}
        SERVICE_DEFS.each { |k, v| @services[k] = v }
      end

      def all
        @mutex.synchronize { @services.values.dup }
      end

      
      def get_service_url(name, host: 'localhost')
        @mutex.synchronize do
          svc = @services[name.to_sym]
          return nil unless svc

          "http://#{host}:#{svc.port}" 
        end
      end

      def validate_contract(name)
        @mutex.synchronize do
          svc = @services[name.to_sym]
          return false unless svc

          svc.dependencies.all? { |dep| @services.key?(dep) }
        end
      end

      
      def topological_order
        @mutex.synchronize do
          in_degree = {}
          adjacency = {}
          @services.each do |k, svc|
            in_degree[k] ||= 0
            adjacency[k] ||= []
            svc.dependencies.each do |dep|
              in_degree[dep] ||= 0
              adjacency[dep] ||= []
              adjacency[dep] << k
              in_degree[k] += 1
            end
          end

          queue = in_degree.select { |_, d| d.zero? }.keys.sort
          result = []
          until queue.empty?
            node = queue.shift
            result << node
            (adjacency[node] || []).each do |neighbor|
              in_degree[neighbor] -= 1
              queue << neighbor if in_degree[neighbor].zero?
            end
            queue.sort!
          end
          
          result
        end
      end

      def service_count
        @mutex.synchronize { @services.length }
      end

      
      def dependency_depth(name)
        @mutex.synchronize do
          svc = @services[name.to_sym]
          return 0 unless svc

          visited = {}
          queue = [[name.to_sym, 0]]
          max_depth = 0
          until queue.empty?
            current, depth = queue.shift
            next if visited[current]

            visited[current] = true
            max_depth = depth if depth > max_depth
            s = @services[current]
            next unless s

            s.dependencies.each { |dep| queue << [dep, depth + 1] }
          end
          max_depth 
        end
      end
    end
  end
end
